File: scripts_archive/node/scripts/parse_pa_catalog.mjs
#!/usr/bin/env node
/**
 * Parse Pennsylvania PLCB Wholesale Catalog Excel → JSON
 * Extracts wine products with UPC barcodes, prices, vintage, country, region.
 * Output: data/imports/pa_wines_parsed.json
 */
import XLSX from 'xlsx';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const INPUT = path.join(__dirname, '..', 'data', 'imports', 'pa_wine_catalog.xlsx');
const OUTPUT = path.join(__dirname, '..', 'data', 'imports', 'pa_wines_parsed.json');

const wb = XLSX.readFile(INPUT);
const ws = wb.Sheets[wb.SheetNames[0]];
const data = XLSX.utils.sheet_to_json(ws, { header: 1 });
const rows = data.slice(1).filter(r => r[0] === 'Stock Wine');

const wineGroups = new Set(['Table Wine', 'Fortified Wine', 'Other-Dessert Wines', 'Sparkling Wine']);
const wines = rows.filter(r => wineGroups.has(r[1]));

const output = wines.map(r => {
  const upcs = [r[16], r[17], r[18], r[19], r[20]]
    .filter(u => u)
    .map(u => String(u).trim());

  const vintage = r[28];
  let vintageYear = null;
  if (vintage && /^\d{4}$/.test(vintage)) vintageYear = parseInt(vintage);

  return {
    plcb_item: r[3] ? String(r[3]).trim() : null,
    name: r[4] || null,
    plcb_scc: r[5] ? String(r[5]).trim() : null,
    manufacturer_scc: r[6] ? String(r[6]).trim() : null,
    volume: r[7] || null,
    case_pack: r[8] ? parseInt(r[8]) : null,
    price_usd: r[9] ? parseFloat(r[9]) : null,
    upcs,
    upc_primary: upcs[0] || null,
    proof: r[27] || null,
    vintage: vintage || null,
    vintage_year: vintageYear,
    brand: r[29] || null,
    import_domestic: r[30] || null,
    country: r[31] || null,
    region: r[32] || null,
    group: r[1] || null,
    class_name: r[2] || null,
  };
});

// Stats
const allUpcs = new Set();
output.forEach(w => w.upcs.forEach(u => allUpcs.add(u)));

const stats = {
  total: output.length,
  has_upc: output.filter(w => w.upc_primary).length,
  unique_upcs: allUpcs.size,
  has_vintage_year: output.filter(w => w.vintage_year).length,
  has_country: output.filter(w => w.country).length,
  has_region: output.filter(w => w.region).length,
  has_brand: output.filter(w => w.brand).length,
  has_proof: output.filter(w => w.proof !== 'N/A' && w.proof != null).length,
  has_price: output.filter(w => w.price_usd).length,
};

console.log('=== PA PLCB Wine Catalog ===');
console.log(`Total wines: ${stats.total}`);
console.log(`Has UPC: ${stats.has_upc} (${(stats.has_upc / stats.total * 100).toFixed(1)}%)`);
console.log(`Unique UPCs: ${stats.unique_upcs}`);
console.log(`Has vintage year: ${stats.has_vintage_year}`);
console.log(`Has country: ${stats.has_country}`);
console.log(`Has region: ${stats.has_region}`);
console.log(`Has brand: ${stats.has_brand}`);
console.log(`Has proof: ${stats.has_proof}`);
console.log(`Has price: ${stats.has_price}`);

// Country breakdown
const countries = {};
output.forEach(w => { if (w.country) countries[w.country] = (countries[w.country] || 0) + 1; });
console.log('\nTop countries:', Object.fromEntries(
  Object.entries(countries).sort((a, b) => b[1] - a[1]).slice(0, 15)
));

// Group breakdown
const groups = {};
output.forEach(w => { groups[w.group] = (groups[w.group] || 0) + 1; });
console.log('Groups:', groups);

// Write output
const file = {
  metadata: {
    source: 'Pennsylvania PLCB Wholesale Catalog',
    file: 'pa_wine_catalog.xlsx',
    extracted_at: new Date().toISOString(),
    stats
  },
  wines: output
};
fs.writeFileSync(OUTPUT, JSON.stringify(file, null, 2));
console.log(`\nSaved to ${OUTPUT}`);
console.log(`File size: ${(fs.statSync(OUTPUT).size / 1024 / 1024).toFixed(1)} MB`);

File: pipeline/load/parse_pa_catalog.py
#!/usr/bin/env python3
"""
Parse Pennsylvania PLCB Wholesale Catalog Excel -> JSON
Extracts wine products with UPC barcodes, prices, vintage, country, region.
Output: data/imports/pa_wines_parsed.json

Usage:
    python -m pipeline.load.parse_pa_catalog
"""

import json
import sys
from collections import Counter
from datetime import datetime, timezone
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

import openpyxl

DATA_DIR = Path(__file__).resolve().parents[2] / "data" / "imports"
INPUT = DATA_DIR / "pa_wine_catalog.xlsx"
OUTPUT = DATA_DIR / "pa_wines_parsed.json"


def main():
    wb = openpyxl.load_workbook(INPUT, read_only=True)
    ws = wb.active
    all_rows = list(ws.iter_rows(values_only=True))
    wb.close()

    # header=1 style: first row is headers, data starts at row 1
    # Filter: col[0] == 'Stock Wine'
    data = [r for r in all_rows[1:] if r[0] == "Stock Wine"]

    wine_groups = {"Table Wine", "Fortified Wine", "Other-Dessert Wines", "Sparkling Wine"}
    wines = [r for r in data if r[1] in wine_groups]

    output = []
    for r in wines:
        upcs = [str(r[i]).strip() for i in range(16, 21) if r[i]]
        vintage = r[28]
        vintage_year = int(vintage) if vintage and str(vintage).strip().isdigit() and len(str(vintage).strip()) == 4 else None

        output.append({
            "plcb_item": str(r[3]).strip() if r[3] else None,
            "name": r[4] or None,
            "plcb_scc": str(r[5]).strip() if r[5] else None,
            "manufacturer_scc": str(r[6]).strip() if r[6] else None,
            "volume": r[7] or None,
            "case_pack": int(r[8]) if r[8] else None,
            "price_usd": float(r[9]) if r[9] else None,
            "upcs": upcs,
            "upc_primary": upcs[0] if upcs else None,
            "proof": r[27] or None,
            "vintage": vintage or None,
            "vintage_year": vintage_year,
            "brand": r[29] or None,
            "import_domestic": r[30] or None,
            "country": r[31] or None,
            "region": r[32] or None,
            "group": r[1] or None,
            "class_name": r[2] or None,
        })

    all_upcs = set()
    for w in output:
        for u in w["upcs"]:
            all_upcs.add(u)

    stats = {
        "total": len(output),
        "has_upc": sum(1 for w in output if w["upc_primary"]),
        "unique_upcs": len(all_upcs),
        "has_vintage_year": sum(1 for w in output if w["vintage_year"]),
        "has_country": sum(1 for w in output if w["country"]),
        "has_region": sum(1 for w in output if w["region"]),
        "has_brand": sum(1 for w in output if w["brand"]),
        "has_proof": sum(1 for w in output if w["proof"] and w["proof"] != "N/A"),
        "has_price": sum(1 for w in output if w["price_usd"]),
    }

    print("=== PA PLCB Wine Catalog ===")
    print(f"Total wines: {stats['total']}")
    print(f"Has UPC: {stats['has_upc']} ({stats['has_upc'] / stats['total'] * 100:.1f}%)")
    print(f"Unique UPCs: {stats['unique_upcs']}")
    print(f"Has vintage year: {stats['has_vintage_year']}")
    print(f"Has country: {stats['has_country']}")
    print(f"Has region: {stats['has_region']}")
    print(f"Has brand: {stats['has_brand']}")
    print(f"Has proof: {stats['has_proof']}")
    print(f"Has price: {stats['has_price']}")

    countries = Counter(w["country"] for w in output if w["country"])
    print("\nTop countries:", dict(countries.most_common(15)))

    groups = Counter(w["group"] for w in output)
    print("Groups:", dict(groups))

    file_data = {
        "metadata": {
            "source": "Pennsylvania PLCB Wholesale Catalog",
            "file": "pa_wine_catalog.xlsx",
            "extracted_at": datetime.now(timezone.utc).isoformat(),
            "stats": stats,
        },
        "wines": output,
    }
    OUTPUT.write_text(json.dumps(file_data, indent=2), encoding="utf-8")
    print(f"\nSaved to {OUTPUT}")
    print(f"File size: {OUTPUT.stat().st_size / (1024 * 1024):.1f} MB")


if __name__ == "__main__":
    main()
